File: services/questradeService.js

import fetch from 'node-fetch';
import fs from 'fs';
import path from 'path';
import { EventEmitter } from 'events';

// Interval mapping: our shorthand -> Questrade format
const INTERVAL_MAP = {
    '1m': 'OneMinute',
    '2m': 'TwoMinutes',
    '3m': 'ThreeMinutes',
    '5m': 'FiveMinutes',
    '10m': 'TenMinutes',
    '15m': 'FifteenMinutes',
    '20m': 'TwentyMinutes',
    '30m': 'HalfHour',
    '1h': 'OneHour',
    '2h': 'TwoHours',
    '4h': 'FourHours',
    '1d': 'OneDay',
    '1w': 'OneWeek',
    '1M': 'OneMonth',
};

/**
 * Questrade API Service
 * Handles OAuth2 authentication, trading endpoints, and market data
 */
export class QuestradeService extends EventEmitter {
    constructor(config = {}) {
        super();
        this.refreshToken = config.refreshToken || process.env.QUESTRADE_REFRESH_TOKEN;
        this.accessToken = null;
        this.apiUrl = null;
        this.isPractice = config.isPractice ?? false;
        this.tokenPath = path.resolve('./questrade_tokens.json');
        this.tokenExpiry = null;

        // Caches
        this.symbolIdCache = new Map(); // ticker -> symbolId
        this.symbolCache = new Map(); // exchange -> symbols[]
        this.symbolCacheTime = new Map(); // exchange -> timestamp

        // Quote polling
        this.pollInterval = null;
        this.activeSubscriptions = [];

        // Try to load saved tokens
        this.loadTokens();
    }

    loadTokens() {
        try {
            if (fs.existsSync(this.tokenPath)) {
                const data = JSON.parse(fs.readFileSync(this.tokenPath, 'utf8'));
                this.refreshToken = data.refreshToken || this.refreshToken;
                this.accessToken = data.accessToken;
                this.apiUrl = data.apiUrl;
                this.tokenExpiry = data.tokenExpiry;
            }
        } catch (e) {
            console.error('[Questrade] Failed to load tokens:', e.message);
        }
    }

    saveTokens(data) {
        try {
            const tokenData = {
                refreshToken: data.refreshToken,
                accessToken: data.accessToken,
                apiUrl: data.apiUrl,
                tokenExpiry: Date.now() + (data.expiresIn * 1000),
            };
            fs.writeFileSync(this.tokenPath, JSON.stringify(tokenData, null, 2));
            this.refreshToken = tokenData.refreshToken;
            this.accessToken = tokenData.accessToken;
            this.apiUrl = tokenData.apiUrl;
            this.tokenExpiry = tokenData.tokenExpiry;
        } catch (e) {
            console.error('[Questrade] Failed to save tokens:', e.message);
        }
    }

    isAuthenticated() {
        const SAFETY_MARGIN = 60 * 1000; // 1 minute before expiry
        return !!(this.accessToken && this.tokenExpiry && Date.now() < (this.tokenExpiry - SAFETY_MARGIN) && this.apiUrl);
    }

    async authenticate(refreshTokenOverride) {
        if (refreshTokenOverride) {
            this.refreshToken = refreshTokenOverride;
        }

        // If we have a valid access token with safety margin, skip
        if (this.isAuthenticated()) {
            return;
        }

        if (!this.refreshToken) {
            throw new Error('Questrade Refresh Token is missing. Please set QUESTRADE_REFRESH_TOKEN in .env or provide it via the API.');
        }

        const loginUrl = this.isPractice
            ? `https://practicelogin.questrade.com/oauth2/token?grant_type=refresh_token&refresh_token=${this.refreshToken}`
            : `https://login.questrade.com/oauth2/token?grant_type=refresh_token&refresh_token=${this.refreshToken}`;

        const response = await fetch(loginUrl, { method: 'POST' });

        if (!response.ok) {
            const err = await response.text();
            throw new Error(`Questrade Authentication Failed: ${err}`);
        }

        const data = await response.json();

        this.saveTokens({
            accessToken: data.access_token,
            refreshToken: data.refresh_token,
            apiUrl: data.api_server,
            expiresIn: data.expires_in
        });

        console.log(`[Questrade] Authenticated (${this.isPractice ? 'Practice' : 'Live'})`);
        this.emit('status', 'authenticated');
    }

    async request(endpoint, method = 'GET', body = null) {
        await this.authenticate();

        const url = `${this.apiUrl}v1/${endpoint}`;
        const options = {
            method,
            headers: {
                'Authorization': `Bearer ${this.accessToken}`,
                'Content-Type': 'application/json'
            }
        };

        if (body) {
            options.body = JSON.stringify(body);
        }

        const response = await fetch(url, options);

        if (!response.ok) {
            const errText = await response.text();
            throw new Error(`Questrade API Error [${endpoint}]: ${response.status} - ${errText}`);
        }

        return response.json();
    }

    // --- Market Data ---

    /**
     * Map our interval shorthand to Questrade format
     */
    mapInterval(interval) {
        return INTERVAL_MAP[interval] || interval;
    }

    /**
     * Normalize Questrade candle format to our internal {o, h, l, c, v, t} format
     */
    normalizeCandles(questradeCandles) {
        if (!questradeCandles || !Array.isArray(questradeCandles)) return [];
        return questradeCandles.map(candle => ({
            t: new Date(candle.start || candle.end).getTime(),
            o: candle.open,
            h: candle.high,
            l: candle.low,
            c: candle.close,
            v: candle.volume,
            vwap: candle.VWAP || candle.vwap || 0,
        }));
    }

    /**
     * Get Candles by symbolId (raw Questrade format)
     */
    async getCandles(symbolId, interval = 'OneMinute', startTime, endTime) {
        const params = new URLSearchParams({
            startTime: startTime || new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
            endTime: endTime || new Date().toISOString(),
            interval: this.mapInterval(interval)
        });
        const data = await this.request(`markets/candles/${symbolId}?${params}`);
        return data.candles || [];
    }

    /**
     * Get Candles by ticker string (convenience method)
     * Resolves ticker to symbolId, then fetches and normalizes candles
     */
    async getCandlesByTicker(ticker, interval = '1m', startTime, endTime) {
        const symbolId = await this.getSymbolId(ticker);
        const rawCandles = await this.getCandles(symbolId, interval, startTime, endTime);
        return this.normalizeCandles(rawCandles);
    }

    /**
     * Get real-time quote for a symbol
     */
    async getQuote(symbolId) {
        const data = await this.request(`markets/quotes/${symbolId}`);
        return data.quotes ? data.quotes[0] : null;
    }

    /**
     * Get quotes for multiple symbols
     */
    async getQuotes(symbolIds) {
        const ids = Array.isArray(symbolIds) ? symbolIds.join(',') : symbolIds;
        const data = await this.request(`markets/quotes?ids=${ids}`);
        return data.quotes || [];
    }

    /**
     * Get quote by ticker string
     */
    async getQuoteByTicker(ticker) {
        const symbolId = await this.getSymbolId(ticker);
        return this.getQuote(symbolId);
    }

    /**
     * Search for a symbol to get its ID
     */
    async searchSymbol(prefix) {
        const data = await this.request(`symbols/search?prefix=${encodeURIComponent(prefix)}`);
        return data.symbols || [];
    }

    async getSymbolId(ticker) {
        if (this.symbolIdCache.has(ticker)) {
            return this.symbolIdCache.get(ticker);
        }

        const symbols = await this.searchSymbol(ticker);
        const match = symbols.find(s => s.symbol === ticker);
        if (!match) throw new Error(`Symbol ${ticker} not found on Questrade`);

        this.symbolIdCache.set(ticker, match.symbolId);
        return match.symbolId;
    }

    async getSymbolIds(tickers) {
        return Promise.all(tickers.map(t => this.getSymbolId(t)));
    }

    /**
     * Get symbols for a specific exchange (cached for 1 hour)
     */
    async getSymbolsByExchange(exchange) {
        const cacheKey = exchange;
        const cachedTime = this.symbolCacheTime.get(cacheKey) || 0;

        if (Date.now() - cachedTime < 3600000 && this.symbolCache.has(cacheKey)) {
            return this.symbolCache.get(cacheKey);
        }

        // Questrade doesn't have a direct "list by exchange" endpoint,
        // so we search with common prefixes and filter by exchange
        const commonPrefixes = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
                                'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];

        // Only search a few popular prefixes to avoid rate limits
        const topPrefixes = ['SH', 'TD', 'RY', 'BN', 'CN', 'CP', 'ENB', 'BMO', 'BB', 'AC'];
        const allSymbols = [];

        for (const prefix of topPrefixes) {
            try {
                const symbols = await this.searchSymbol(prefix);
                const filtered = symbols.filter(s =>
                    s.listingExchange === exchange && s.securityType === 'Stock'
                );
                allSymbols.push(...filtered);
            } catch (e) {
                // Skip failed searches
            }
        }

        // Dedupe by symbolId
        const seen = new Set();
        const unique = allSymbols.filter(s => {
            if (seen.has(s.symbolId)) return false;
            seen.add(s.symbolId);
            return true;
        });

        this.symbolCache.set(cacheKey, unique);
        this.symbolCacheTime.set(cacheKey, Date.now());
        return unique;
    }

    // --- Account & Trading ---

    async getAccounts() {
        const data = await this.request('accounts');
        return data.accounts || [];
    }

    async getBalance(accountId) {
        const data = await this.request(`accounts/${accountId}/balances`);
        return data;
    }

    async getPositions(accountId) {
        const data = await this.request(`accounts/${accountId}/positions`);
        return data.positions || [];
    }

    async placeOrder(accountId, order) {
        return this.request(`accounts/${accountId}/orders`, 'POST', order);
    }

    async getOrders(accountId) {
        const data = await this.request(`accounts/${accountId}/orders`);
        return data.orders || [];
    }

    // --- Real-time Quotes (Polling) ---

    async subscribeQuotes(symbolIds) {
        this.activeSubscriptions = symbolIds;

        if (this.pollInterval) clearInterval(this.pollInterval);

        this.pollInterval = setInterval(async () => {
            try {
                const quotes = await this.getQuotes(symbolIds);
                quotes.forEach(quote => {
                    this.emit('quote', quote);
                });
            } catch (e) {
                console.error('[Questrade] Quote polling error:', e.message);
            }
        }, 1000);
    }

    /**
     * Get connection status
     */
    getStatus() {
        return {
            isAuthenticated: this.isAuthenticated(),
            isPractice: this.isPractice,
            apiUrl: this.apiUrl,
            tokenExpiry: this.tokenExpiry,
            symbolsCached: this.symbolIdCache.size,
        };
    }

    close() {
        if (this.pollInterval) clearInterval(this.pollInterval);
    }
}

File: canuck-trader-pro/backend/services/questrade_service.py
"""
Questrade API Service
Port of services/questradeService.js.

Handles OAuth2 authentication, trading endpoints, and market data for Questrade.
"""

import json
import logging
import os
import time
from pathlib import Path
from datetime import datetime, timedelta, timezone

import httpx

logger = logging.getLogger("questrade")

INTERVAL_MAP = {
    "1m": "OneMinute", "2m": "TwoMinutes", "3m": "ThreeMinutes",
    "5m": "FiveMinutes", "10m": "TenMinutes", "15m": "FifteenMinutes",
    "20m": "TwentyMinutes", "30m": "HalfHour", "1h": "OneHour",
    "2h": "TwoHours", "4h": "FourHours", "1d": "OneDay",
    "1w": "OneWeek", "1M": "OneMonth",
}


class QuestradeService:
    def __init__(self, is_practice: bool = False):
        self.refresh_token = os.environ.get("QUESTRADE_REFRESH_TOKEN", "")
        self.access_token: str | None = None
        self.api_url: str | None = None
        self.is_practice = is_practice
        self.token_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "questrade_tokens.json")
        self.token_expiry: float = 0

        self.symbol_id_cache: dict[str, int] = {}
        self.symbol_cache: dict[str, list] = {}
        self.symbol_cache_time: dict[str, float] = {}

        self._load_tokens()

    def _load_tokens(self):
        try:
            if os.path.exists(self.token_path):
                with open(self.token_path) as f:
                    data = json.load(f)
                self.refresh_token = data.get("refresh_token", self.refresh_token)
                self.access_token = data.get("access_token")
                self.api_url = data.get("api_url")
                self.token_expiry = data.get("token_expiry", 0)
        except Exception as e:
            logger.error(f"Failed to load tokens: {e}")

    def _save_tokens(self, data: dict):
        try:
            token_data = {
                "refresh_token": data["refresh_token"],
                "access_token": data["access_token"],
                "api_url": data["api_url"],
                "token_expiry": time.time() + data["expires_in"],
            }
            with open(self.token_path, "w") as f:
                json.dump(token_data, f, indent=2)
            self.refresh_token = token_data["refresh_token"]
            self.access_token = token_data["access_token"]
            self.api_url = token_data["api_url"]
            self.token_expiry = token_data["token_expiry"]
        except Exception as e:
            logger.error(f"Failed to save tokens: {e}")

    def is_authenticated(self) -> bool:
        return bool(
            self.access_token
            and self.token_expiry
            and time.time() < self.token_expiry - 60
            and self.api_url
        )

    async def authenticate(self, refresh_token_override: str | None = None):
        if refresh_token_override:
            self.refresh_token = refresh_token_override

        if self.is_authenticated():
            return

        if not self.refresh_token:
            raise ValueError("Questrade Refresh Token is missing.")

        base = "https://practicelogin.questrade.com" if self.is_practice else "https://login.questrade.com"
        url = f"{base}/oauth2/token?grant_type=refresh_token&refresh_token={self.refresh_token}"

        async with httpx.AsyncClient() as client:
            resp = await client.post(url)
            if resp.status_code != 200:
                raise RuntimeError(f"Questrade auth failed: {resp.text}")
            data = resp.json()

        self._save_tokens({
            "access_token": data["access_token"],
            "refresh_token": data["refresh_token"],
            "api_url": data["api_server"],
            "expires_in": data["expires_in"],
        })
        logger.info(f"Authenticated ({'Practice' if self.is_practice else 'Live'})")

    async def request(self, endpoint: str, method: str = "GET", body: dict | None = None) -> dict:
        await self.authenticate()
        url = f"{self.api_url}v1/{endpoint}"
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }
        async with httpx.AsyncClient() as client:
            if method == "GET":
                resp = await client.get(url, headers=headers)
            else:
                resp = await client.request(method, url, headers=headers, json=body)

        if resp.status_code != 200:
            raise RuntimeError(f"Questrade API Error [{endpoint}]: {resp.status_code} - {resp.text}")
        return resp.json()

    def map_interval(self, interval: str) -> str:
        return INTERVAL_MAP.get(interval, interval)

    def normalize_candles(self, raw_candles: list[dict]) -> list[dict]:
        if not raw_candles:
            return []
        result = []
        for c in raw_candles:
            ts = c.get("start") or c.get("end") or ""
            try:
                t = int(datetime.fromisoformat(ts.replace("Z", "+00:00")).timestamp() * 1000)
            except Exception:
                t = 0
            result.append({
                "t": t, "o": c.get("open", 0), "h": c.get("high", 0),
                "l": c.get("low", 0), "c": c.get("close", 0),
                "v": c.get("volume", 0), "vwap": c.get("VWAP") or c.get("vwap") or 0,
            })
        return result

    async def get_candles(self, symbol_id: int, interval: str = "OneMinute",
                          start_time: str | None = None, end_time: str | None = None) -> list[dict]:
        if not start_time:
            start_time = (datetime.now(timezone.utc) - timedelta(days=1)).isoformat()
        if not end_time:
            end_time = datetime.now(timezone.utc).isoformat()
        endpoint = f"markets/candles/{symbol_id}?startTime={start_time}&endTime={end_time}&interval={self.map_interval(interval)}"
        data = await self.request(endpoint)
        return data.get("candles", [])

    async def get_candles_by_ticker(self, ticker: str, interval: str = "1m",
                                     start_time: str | None = None, end_time: str | None = None) -> list[dict]:
        sid = await self.get_symbol_id(ticker)
        raw = await self.get_candles(sid, interval, start_time, end_time)
        return self.normalize_candles(raw)

    async def get_quote(self, symbol_id: int) -> dict | None:
        data = await self.request(f"markets/quotes/{symbol_id}")
        quotes = data.get("quotes", [])
        return quotes[0] if quotes else None

    async def get_quotes(self, symbol_ids: list[int]) -> list[dict]:
        ids = ",".join(str(s) for s in symbol_ids)
        data = await self.request(f"markets/quotes?ids={ids}")
        return data.get("quotes", [])

    async def search_symbol(self, prefix: str) -> list[dict]:
        data = await self.request(f"symbols/search?prefix={prefix}")
        return data.get("symbols", [])

    async def get_symbol_id(self, ticker: str) -> int:
        if ticker in self.symbol_id_cache:
            return self.symbol_id_cache[ticker]
        symbols = await self.search_symbol(ticker)
        match = next((s for s in symbols if s["symbol"] == ticker), None)
        if not match:
            raise ValueError(f"Symbol {ticker} not found on Questrade")
        self.symbol_id_cache[ticker] = match["symbolId"]
        return match["symbolId"]

    async def get_accounts(self) -> list[dict]:
        data = await self.request("accounts")
        return data.get("accounts", [])

    async def get_balance(self, account_id: str) -> dict:
        data = await self.request(f"accounts/{account_id}/balances")
        return data

    async def get_positions(self, account_id: str) -> list[dict]:
        data = await self.request(f"accounts/{account_id}/positions")
        return data.get("positions", [])

    async def place_order(self, account_id: str, order: dict) -> dict:
        return await self.request(f"accounts/{account_id}/orders", method="POST", body=order)

    def get_status(self) -> dict:
        return {
            "authenticated": self.is_authenticated(),
            "is_practice": self.is_practice,
            "api_url": self.api_url or "",
            "token_expiry": self.token_expiry,
            "cached_symbols": len(self.symbol_id_cache),
        }
